Make Listings vector lazy

```rust 
fn listings_iter<'a>(&'a mut self, site: &'a RedditSite) -> Box<dyn Iterator<Item=Comments> + 'a>{
            let comments: Box<dyn Iterator<Item=Comments>> = Box::new(site.data.children
                .iter()
                .map(|post| {
                    let url = format!("{}{}.json", &mut self.base_url, post.data.permalink);
                    let data = self.webclient.data(&url);

                    Comments::new(&data.clone())
                }).into_iter());
            comments
    }
```

```rust
error[E0373]: closure may outlive the current function, but it borrows `self`, which is owned by the current function
  --> src/reddit/lib.rs:62:22
   |
62 |                 .map(|post| {
   |                      ^^^^^^ may outlive borrowed value `self`
...
65 |                     let url = format!("{}{}.json", &mut self.base_url, post.data.permalink);
   |                                                         ---- `self` is borrowed here
   |
```


https://users.rust-lang.org/t/referencing-self-from-within-closure-in-a-method/19189/2

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -4,13 +4,11 @@ use reddit::RedditClient;
 
 
 pub fn run() {
-    let listings = RedditClient::new("/r/all").limit(3).get();
-
-    for listing in listings {
+    for listing in RedditClient::new("/r/all").limit(3) {
         println!("\n\n'{}' has {} comments.\n\t\thttps://reddit.com{}\n\t ",
                  listing.link.data.children[0].data.title,
                  listing.comments.data.children.len(),
                  listing.link.data.children[0].data.permalink,
         );
-    }
+    };
 }

--- a/src/main.rs
+++ b/src/main.rs
@@ -1,4 +1,3 @@
-#[macro_use]
 extern crate log;
 extern crate simple_logger;
 

--- a/src/reddit/lib.rs
+++ b/src/reddit/lib.rs
@@ -7,7 +7,30 @@ pub struct RedditClient {
     subreddit: String,
     queries: String,
     webclient: requests::Cacher,
+    position: usize,
+}
 
+impl Iterator for RedditClient {
+    type Item = Comments;
+    fn next(&mut self) -> Option<Self::Item> {
+        let url = format!("{}{}.json{}", self.base_url, self.subreddit, self.queries);
+        let data = self.webclient.data(&url);
+        let site = RedditSite::new(&data);
+
+        if self.position >= site.data.children.len() {
+            return None
+        };
+
+        let url = format!("{}{}.json",
+                          self.base_url,
+                          site.data.children[self.position].data.permalink
+        );
+        let listing = Comments::new(
+            &self.webclient.data(&url)
+        );
+        self.position += 1;
+        Some(listing)
+    }
 }
 
 
@@ -21,8 +44,10 @@ impl RedditClient {
                 log::warn!("Requesting [GET] {}", url);
                 ureq::get(url).call().into_string().unwrap()
             }),
+            position: 0
         }
     }
+
 
     fn add_query<'a>(&'a mut self, name: String, value: String) -> &'a mut Self {
         self.queries = match self.queries.as_str() {
@@ -36,25 +61,5 @@ impl RedditClient {
         self.add_query(String::from("limit"), format!("{}", value))
     }
 
-    fn download_reddit_site(&mut self) -> RedditSite {
-        let url = format!("{}{}.json{}", self.base_url, self.subreddit, self.queries);
-        let data = self.webclient.data(&url);
-        RedditSite::new(&data)
-    }
 
-    fn download_listings(&mut self, site: RedditSite) -> Vec<Comments> {
-        let mut listings = vec![];
-        for post in site.data.children {
-            let url = format!("{}{}.json", self.base_url, post.data.permalink);
-            listings.push(Comments::new(
-                &self.webclient.data(&url)
-            ));
-        }
-        listings
-    }
-
-    pub fn get<'a>(&'a mut self) -> Vec<Comments> {
-        let reddit_site = self.download_reddit_site();
-        self.download_listings(reddit_site)
-    }
 }

